File: src/com/l2kt/gameserver/network/clientpackets/RequestRefine.java
package com.l2kt.gameserver.network.clientpackets;

import com.l2kt.gameserver.data.xml.AugmentationData;
import com.l2kt.gameserver.model.L2Augmentation;
import com.l2kt.gameserver.model.actor.instance.Player;
import com.l2kt.gameserver.model.item.instance.ItemInstance;
import com.l2kt.gameserver.network.SystemMessageId;
import com.l2kt.gameserver.network.serverpackets.ExVariationResult;
import com.l2kt.gameserver.network.serverpackets.InventoryUpdate;
import com.l2kt.gameserver.network.serverpackets.StatusUpdate;

/**
 * Format:(ch) dddd
 * @author -Wooden-
 */
public final class RequestRefine extends AbstractRefinePacket
{
	private int _targetItemObjId;
	private int _refinerItemObjId;
	private int _gemStoneItemObjId;
	private int _gemStoneCount;
	
	@Override
	protected void readImpl()
	{
		_targetItemObjId = readD();
		_refinerItemObjId = readD();
		_gemStoneItemObjId = readD();
		_gemStoneCount = readD();
	}
	
	@Override
	protected void runImpl()
	{
		final Player activeChar = getClient().getActiveChar();
		if (activeChar == null)
			return;
		
		final ItemInstance targetItem = activeChar.getInventory().getItemByObjectId(_targetItemObjId);
		if (targetItem == null)
		{
			activeChar.sendPacket(new ExVariationResult(0, 0, 0));
			activeChar.sendPacket(SystemMessageId.AUGMENTATION_FAILED_DUE_TO_INAPPROPRIATE_CONDITIONS);
			return;
		}
		
		final ItemInstance refinerItem = activeChar.getInventory().getItemByObjectId(_refinerItemObjId);
		if (refinerItem == null)
		{
			activeChar.sendPacket(new ExVariationResult(0, 0, 0));
			activeChar.sendPacket(SystemMessageId.AUGMENTATION_FAILED_DUE_TO_INAPPROPRIATE_CONDITIONS);
			return;
		}
		
		final ItemInstance gemStoneItem = activeChar.getInventory().getItemByObjectId(_gemStoneItemObjId);
		if (gemStoneItem == null)
		{
			activeChar.sendPacket(new ExVariationResult(0, 0, 0));
			activeChar.sendPacket(SystemMessageId.AUGMENTATION_FAILED_DUE_TO_INAPPROPRIATE_CONDITIONS);
			return;
		}
		
		if (!Companion.isValid(activeChar, targetItem, refinerItem, gemStoneItem))
		{
			activeChar.sendPacket(new ExVariationResult(0, 0, 0));
			activeChar.sendPacket(SystemMessageId.AUGMENTATION_FAILED_DUE_TO_INAPPROPRIATE_CONDITIONS);
			return;
		}
		
		final LifeStone ls = Companion.getLifeStone(refinerItem.getItemId());
		if (ls == null)
		{
			activeChar.sendPacket(new ExVariationResult(0, 0, 0));
			activeChar.sendPacket(SystemMessageId.AUGMENTATION_FAILED_DUE_TO_INAPPROPRIATE_CONDITIONS);
			return;
		}
		
		final int lifeStoneLevel = ls.getLevel();
		final int lifeStoneGrade = ls.getGrade();
		if (_gemStoneCount != Companion.getGemStoneCount(targetItem.getItem().getCrystalType()))
		{
			activeChar.sendPacket(new ExVariationResult(0, 0, 0));
			activeChar.sendPacket(SystemMessageId.AUGMENTATION_FAILED_DUE_TO_INAPPROPRIATE_CONDITIONS);
			return;
		}
		
		// unequip item
		if (targetItem.isEquipped())
		{
			ItemInstance[] unequipped = activeChar.getInventory().unEquipItemInSlotAndRecord(targetItem.getLocationSlot());
			InventoryUpdate iu = new InventoryUpdate();
			
			for (ItemInstance itm : unequipped)
				iu.addModifiedItem(itm);
			
			activeChar.sendPacket(iu);
			activeChar.broadcastUserInfo();
		}
		
		// Consume the life stone
		if (!activeChar.destroyItem("RequestRefine", refinerItem, 1, null, false))
		{
			activeChar.sendPacket(new ExVariationResult(0, 0, 0));
			activeChar.sendPacket(SystemMessageId.AUGMENTATION_FAILED_DUE_TO_INAPPROPRIATE_CONDITIONS);
			return;
		}
		
		// Consume gemstones
		if (!activeChar.destroyItem("RequestRefine", gemStoneItem, _gemStoneCount, null, false))
		{
			activeChar.sendPacket(new ExVariationResult(0, 0, 0));
			activeChar.sendPacket(SystemMessageId.AUGMENTATION_FAILED_DUE_TO_INAPPROPRIATE_CONDITIONS);
			return;
		}
		
		final L2Augmentation aug = AugmentationData.getInstance().generateRandomAugmentation(lifeStoneLevel, lifeStoneGrade);
		targetItem.setAugmentation(aug);
		
		final int stat12 = 0x0000FFFF & aug.getAugmentationId();
		final int stat34 = aug.getAugmentationId() >> 16;
		activeChar.sendPacket(new ExVariationResult(stat12, stat34, 1));
		
		InventoryUpdate iu = new InventoryUpdate();
		iu.addModifiedItem(targetItem);
		activeChar.sendPacket(iu);
		
		StatusUpdate su = new StatusUpdate(activeChar);
		su.addAttribute(StatusUpdate.CUR_LOAD, activeChar.getCurrentLoad());
		activeChar.sendPacket(su);
	}
}

File: src/com/l2kt/gameserver/network/clientpackets/RequestRefine.kt
package com.l2kt.gameserver.network.clientpackets

import com.l2kt.gameserver.data.xml.AugmentationData
import com.l2kt.gameserver.network.SystemMessageId
import com.l2kt.gameserver.network.serverpackets.ExVariationResult
import com.l2kt.gameserver.network.serverpackets.InventoryUpdate
import com.l2kt.gameserver.network.serverpackets.StatusUpdate

/**
 * Format:(ch) dddd
 * @author -Wooden-
 */
class RequestRefine : AbstractRefinePacket() {
    private var _targetItemObjId: Int = 0
    private var _refinerItemObjId: Int = 0
    private var _gemStoneItemObjId: Int = 0
    private var _gemStoneCount: Int = 0

    override fun readImpl() {
        _targetItemObjId = readD()
        _refinerItemObjId = readD()
        _gemStoneItemObjId = readD()
        _gemStoneCount = readD()
    }

    override fun runImpl() {
        val activeChar = client.activeChar ?: return

        val targetItem = activeChar.inventory!!.getItemByObjectId(_targetItemObjId)
        if (targetItem == null) {
            activeChar.sendPacket(ExVariationResult(0, 0, 0))
            activeChar.sendPacket(SystemMessageId.AUGMENTATION_FAILED_DUE_TO_INAPPROPRIATE_CONDITIONS)
            return
        }

        val refinerItem = activeChar.inventory!!.getItemByObjectId(_refinerItemObjId)
        if (refinerItem == null) {
            activeChar.sendPacket(ExVariationResult(0, 0, 0))
            activeChar.sendPacket(SystemMessageId.AUGMENTATION_FAILED_DUE_TO_INAPPROPRIATE_CONDITIONS)
            return
        }

        val gemStoneItem = activeChar.inventory!!.getItemByObjectId(_gemStoneItemObjId)
        if (gemStoneItem == null) {
            activeChar.sendPacket(ExVariationResult(0, 0, 0))
            activeChar.sendPacket(SystemMessageId.AUGMENTATION_FAILED_DUE_TO_INAPPROPRIATE_CONDITIONS)
            return
        }

        if (!AbstractRefinePacket.isValid(activeChar, targetItem, refinerItem, gemStoneItem)) {
            activeChar.sendPacket(ExVariationResult(0, 0, 0))
            activeChar.sendPacket(SystemMessageId.AUGMENTATION_FAILED_DUE_TO_INAPPROPRIATE_CONDITIONS)
            return
        }

        val ls = AbstractRefinePacket.getLifeStone(refinerItem.itemId)
        if (ls == null) {
            activeChar.sendPacket(ExVariationResult(0, 0, 0))
            activeChar.sendPacket(SystemMessageId.AUGMENTATION_FAILED_DUE_TO_INAPPROPRIATE_CONDITIONS)
            return
        }

        val lifeStoneLevel = ls.level
        val lifeStoneGrade = ls.grade
        if (_gemStoneCount != AbstractRefinePacket.getGemStoneCount(targetItem.item.crystalType)) {
            activeChar.sendPacket(ExVariationResult(0, 0, 0))
            activeChar.sendPacket(SystemMessageId.AUGMENTATION_FAILED_DUE_TO_INAPPROPRIATE_CONDITIONS)
            return
        }

        // unequip item
        if (targetItem.isEquipped) {
            val unequipped = activeChar.inventory!!.unEquipItemInSlotAndRecord(targetItem.locationSlot)
            val iu = InventoryUpdate()

            for (itm in unequipped)
                iu.addModifiedItem(itm)

            activeChar.sendPacket(iu)
            activeChar.broadcastUserInfo()
        }

        // Consume the life stone
        if (!activeChar.destroyItem("RequestRefine", refinerItem, 1, null, false)) {
            activeChar.sendPacket(ExVariationResult(0, 0, 0))
            activeChar.sendPacket(SystemMessageId.AUGMENTATION_FAILED_DUE_TO_INAPPROPRIATE_CONDITIONS)
            return
        }

        // Consume gemstones
        if (!activeChar.destroyItem("RequestRefine", gemStoneItem, _gemStoneCount, null, false)) {
            activeChar.sendPacket(ExVariationResult(0, 0, 0))
            activeChar.sendPacket(SystemMessageId.AUGMENTATION_FAILED_DUE_TO_INAPPROPRIATE_CONDITIONS)
            return
        }

        val aug = AugmentationData.generateRandomAugmentation(lifeStoneLevel, lifeStoneGrade)
        targetItem.augmentation = aug

        val stat12 = 0x0000FFFF and aug.augmentationId
        val stat34 = aug.augmentationId shr 16
        activeChar.sendPacket(ExVariationResult(stat12, stat34, 1))

        val iu = InventoryUpdate()
        iu.addModifiedItem(targetItem)
        activeChar.sendPacket(iu)

        val su = StatusUpdate(activeChar)
        su.addAttribute(StatusUpdate.CUR_LOAD, activeChar.currentLoad)
        activeChar.sendPacket(su)
    }
}